resources/aws_ec2.go

```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
 	return AwsPrefix + obj.GetName()
 }
 
+// longpollRunningWaiter waits for the instance to stop and waits for it to
+// terminate. If either waiter returns, the instance state is checked, and an
+// awsEc2Event is returned.
+func longpollRunningWaiter(ctx context.Context, waitInput *ec2.DescribeInstancesInput, c *ec2.EC2) (awsEc2Event, error) {
+	if err := waitUntilInstanceStoppedOrTerminatedWithContext(ctx, waitInput, c); err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error waiting for instance to stop or terminate")
+	}
+	// Check the instance state, and return the appropriate event.
+	stateOutput, err := c.DescribeInstances(waitInput)
+	if err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error describing instances")
+	}
+	if len(stateOutput.Reservations) == 1 {
+		switch *stateOutput.Reservations[0].Instances[0].State.Name {
+		case "stopped":
+			return awsEc2EventInstanceStopped, nil
+		case "terminated":
+			return awsEc2EventInstanceTerminated, nil
+		}
+	}
+	return awsEc2EventNone, nil
+}
+
+// waitUntilInstanceStoppedOrTerminatedWithContext combines the two waiters
+// required to trigger events if a running instance is stopped or terminated.
+// This function is needed, because the AWS api only provides waiters for one
+// or the other.
+func waitUntilInstanceStoppedOrTerminatedWithContext(ctx context.Context, waitInput *ec2.DescribeInstancesInput, c *ec2.EC2) error {
+	errChan := make(chan error)
+	defer close(errChan)
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+	closeChan := make(chan struct{})
+	wgCtx, wgCancel := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-wgCtx.Done():
+			close(closeChan)
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
+		if err := c.WaitUntilInstanceStoppedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to stop"):
+			case <-closeChan:
+			}
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
+		if err := c.WaitUntilInstanceTerminatedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to terminate"):
+			case <-closeChan:
+			}
+		}
+	}()
+	select {
+	case err, ok := <-errChan:
+		if !ok {
+			return fmt.Errorf("error reading from errChan")
```

Comment:
this is not what this means-- rephrase your error message please and explain why you want this scenario to be an error. I'm okay with this, but you have to tell me why.